modern/org/commcare/modern/database/DatabaseHelper.java

```diff
@@ -0,0 +1,153 @@
+package org.commcare.modern.database;
+
+import org.commcare.modern.models.EncryptedModel;
+import org.commcare.modern.models.RecordTooLargeException;
+import org.commcare.modern.util.Pair;
+import org.javarosa.core.services.storage.IMetaData;
+import org.javarosa.core.services.storage.Persistable;
+import org.javarosa.core.services.storage.StorageFullException;
+import org.javarosa.core.util.externalizable.Externalizable;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Setup of platform-agnostic DB helper functions IE for generating SQL
+ * statements, args, content values, etc.
+ *
+ * @author wspride
+ *
+ */
+public class DatabaseHelper {
+
+    public static String ID_COL = "commcare_sql_id";
+    public static String DATA_COL = "commcare_sql_record";
+
+    public static Pair<String, String[]> createWhere(String[] fieldNames, Object[] values,  Persistable p)  throws IllegalArgumentException {
+        return createWhere(fieldNames, values, null, p);
+    }
+
+    public static Pair<String, String[]> createWhere(String[] fieldNames, Object[] values,  EncryptedModel em, Persistable p)  throws IllegalArgumentException {
+        Set<String> fields = null;
+        if(p instanceof IMetaData) {
+            IMetaData m = (IMetaData)p;
+            String[] thefields = m.getMetaDataFields();
+            fields = new HashSet<String>();
+            for(String s : thefields) {
+                fields.add(TableBuilder.scrubName(s));
+            }
+        }
+
+
+        if(em instanceof IMetaData) {
+            IMetaData m = (IMetaData)em;
+            String[] thefields = m.getMetaDataFields();
+            fields = new HashSet<String>();
+            for(String s : thefields) {
+                fields.add(TableBuilder.scrubName(s));
+            }
+        }
+
+
+        String ret = "";
+        ArrayList<String> arguments = new ArrayList<String>();
+        boolean set = false;
+        for(int i = 0 ; i < fieldNames.length; ++i) {
+            String columnName = TableBuilder.scrubName(fieldNames[i]);
+            if(fields != null) {
+                if(!fields.contains(columnName)) {
+                    continue;
+                }
+            }
+
+            if(set){
+                ret += " AND ";
+            }
+
+            ret += columnName + "=?";
```

Comment:
Usually a good idea to use a `StringBuilder` when doing string construction inside of a loop.